### PR TITLE
[bluetooth] Refactor and unify BluetoothAdapter implementation logic

### PR DESCRIPTION
**Pull request by cpmeister**
The goal of these changes is to unify the behavior of inactive device cleanup among the different BluetoothAdapter implementations. Additional code similarities have been abstracted out as well.

There is a breaking change I made to the adapter configurations. I changed the one of the config param names, but since it is typically disabled by default I think the name change will cause a negligible impact.

### CODE REVIEW

**Comment by cpmeister** [comment]
Looks like some wemo test failed. Completely unrelated to the changes in this PR.

**Comment by cpmeister** [comment]
This also fixes a bug where bluegiga doesn't delete discoveries.

**Comment by cpmeister** [comment]
@J-N-K When can we get this merged? I have other bluetooth developers that want to develop support for new types of bluetooth adapters and this needs to be merged before they can start on that.

File: bundles/org.openhab.binding.bluetooth.bluegiga/src/main/resources/ESH-INF/thing/bluegiga.xml [hunk 1]
```diff
@@ -14,12 +14,24 @@
...
-			<parameter name="discovery" type="boolean">
-				<label>Device Discovery</label>
-				<description>Whether this adapter actively participates in Bluetooth device discovery</description>
+			<parameter name="backgroundDiscovery" type="boolean">
+				<label>Background Discovery</label>
+				<description>Whether this adapter performs background discovery of Bluetooth devices</description>
 				<advanced>true</advanced>
 				<default>false</default>
 			</parameter>
+			<parameter name="inactiveDeviceCleanupIntervalSecs" type="integer" min="1">
```
Comment (J-N-K) [inline bundles/org.openhab.binding.bluetooth.bluegiga/src/main/resources/ESH-INF/thing/bluegiga.xml:23]:
```suggestion
			<parameter name="inactiveDeviceCleanupInterval" type="integer" min="1" unit="s">
```

File: bundles/org.openhab.binding.bluetooth.bluegiga/src/main/resources/ESH-INF/thing/bluegiga.xml [hunk 2]
```diff
@@ -14,12 +14,24 @@
...
+				<label>Device Cleanup Interval</label>
+				<description>How often device cleanup is performed in seconds</description>
+				<advanced>true</advanced>
+				<default>60</default>
+			</parameter>
+			<parameter name="inactiveDeviceCleanupThresholdSecs" type="integer" min="1">
```
Comment (J-N-K) [inline bundles/org.openhab.binding.bluetooth.bluegiga/src/main/resources/ESH-INF/thing/bluegiga.xml:29]:
```suggestion
			<parameter name="inactiveDeviceCleanupThreshold" type="integer" min="1" unit="s">
```

File: bundles/org.openhab.binding.bluetooth.bluegiga/src/main/resources/ESH-INF/thing/bluegiga.xml [hunk 3]
```diff
@@ -14,12 +14,24 @@
...
+				<label>Device Cleanup Threshold</label>
+				<description>The number of seconds a device can remain radio silent before it is eligible for cleanup</description>
```
Comment (J-N-K) [inline bundles/org.openhab.binding.bluetooth.bluegiga/src/main/resources/ESH-INF/thing/bluegiga.xml:31]:
```suggestion
				<description>Timespan a device can remain radio silent before it is eligible for cleanup</description>
```

File: bundles/org.openhab.binding.bluetooth.bluegiga/src/main/resources/ESH-INF/thing/bluegiga.xml [hunk 4]
```diff
@@ -14,12 +14,24 @@
...
-				<description>Whether this adapter actively participates in Bluetooth device discovery</description>
+			<parameter name="backgroundDiscovery" type="boolean">
+				<label>Background Discovery</label>
+				<description>Whether this adapter performs background discovery of Bluetooth devices</description>
 				<advanced>true</advanced>
 				<default>false</default>
 			</parameter>
+			<parameter name="inactiveDeviceCleanupIntervalSecs" type="integer" min="1">
+				<label>Device Cleanup Interval</label>
+				<description>How often device cleanup is performed in seconds</description>
```
Comment (J-N-K) [inline bundles/org.openhab.binding.bluetooth.bluegiga/src/main/resources/ESH-INF/thing/bluegiga.xml:25]:
```suggestion
				<description>How often device cleanup is performed</description>
```

File: bundles/org.openhab.binding.bluetooth.bluez/src/main/java/org/openhab/binding/bluetooth/bluez/handler/BlueZBridgeHandler.java [hunk 5]
```diff
@@ -167,70 +112,36 @@ private void startDiscovery() {
...
             startDiscovery();
-            // everything went fine, so lets switch to online
-            updateStatus(ThingStatus.ONLINE);
         } catch (BluetoothException ex) {
             String message = ex.getMessage();
             if (message != null) {
                 if (message.contains("Operation already in progress")) {
                     // we shouldn't go offline in this case
-                    return;
+                    break refreshTry;
```
Comment (J-N-K) [inline bundles/org.openhab.binding.bluetooth.bluez/src/main/java/org/openhab/binding/bluetooth/bluez/handler/BlueZBridgeHandler.java:131]: Nice. Didn't know about that.

File: bundles/org.openhab.binding.bluetooth.bluez/src/main/java/org/openhab/binding/bluetooth/bluez/handler/BlueZBridgeHandler.java [hunk 6]
```diff
@@ -167,70 +112,36 @@ private void startDiscovery() {
...
-    private boolean shouldRemove(BlueZBluetoothDevice device) {
-        // we can't remove devices with listeners since that means they have a handler.
-        if (device.hasListeners()) {
-            return false;
-        }
-        // devices that are connected won't receive any scan notifications so we can't remove them for being idle
-        if (device.getConnectionState() == ConnectionState.CONNECTED) {
-            return false;
+        // everything went fine, so lets switch to online if not already
+        if (getThing().getStatus() != ThingStatus.ONLINE) {
```
Comment (J-N-K) [inline bundles/org.openhab.binding.bluetooth.bluez/src/main/java/org/openhab/binding/bluetooth/bluez/handler/BlueZBridgeHandler.java:142]: I think you can omit the check and let the framework do that work.

File: bundles/org.openhab.binding.bluetooth.bluez/src/main/resources/ESH-INF/thing/bluez.xml [hunk 7]
```diff
@@ -14,12 +14,24 @@
...
-			<parameter name="discovery" type="boolean">
-				<label>Device Discovery</label>
-				<description>Whether Bluetooth device discovery is always enabled for this adapter</description>
+			<parameter name="backgroundDiscovery" type="boolean">
+				<label>Background Discovery</label>
+				<description>Whether this adapter performs background discovery of Bluetooth devices</description>
 				<advanced>true</advanced>
 				<default>false</default>
 			</parameter>
+			<parameter name="inactiveDeviceCleanupIntervalSecs" type="integer" min="1">
```
Comment (J-N-K) [inline bundles/org.openhab.binding.bluetooth.bluez/src/main/resources/ESH-INF/thing/bluez.xml:23]: see above

File: bundles/org.openhab.binding.bluetooth.bluez/src/main/resources/ESH-INF/thing/bluez.xml [hunk 8]
```diff
@@ -14,12 +14,24 @@
...
+				<label>Device Cleanup Interval</label>
+				<description>How often device cleanup is performed in seconds</description>
+				<advanced>true</advanced>
+				<default>60</default>
+			</parameter>
+			<parameter name="inactiveDeviceCleanupThresholdSecs" type="integer" min="1">
```
Comment (J-N-K) [inline bundles/org.openhab.binding.bluetooth.bluez/src/main/resources/ESH-INF/thing/bluez.xml:29]: see above

File: bundles/org.openhab.binding.bluetooth/src/main/java/org/openhab/binding/bluetooth/AbstractBluetoothBridgeHandler.java [hunk 9]
```diff
@@ -0,0 +1,223 @@
...
+    @Override
+    public void scanStop() {
+        // Set active discovery state back to the configured discovery state.
+        activeScanEnabled = false;
+        // We need to keep the adapter in discovery mode as we otherwise won't get any RSSI updates either
+    }
+
+    @Override
+    public BD getDevice(BluetoothAddress address) {
+        synchronized (devices) {
```
Comment (J-N-K) [inline bundles/org.openhab.binding.bluetooth/src/main/java/org/openhab/binding/bluetooth/AbstractBluetoothBridgeHandler.java:173]: why not use `computeIfAbsent` here? `ConcurrentHashMap` implements that atomically.

Comment (cpmeister) [reply bundles/org.openhab.binding.bluetooth/src/main/java/org/openhab/binding/bluetooth/AbstractBluetoothBridgeHandler.java:173]: I toyed with the idea, but the atomic behavior is already handled by the synchronized blocks and computeIfAbsent saves at most 2 lines of code while at the same having to utilize the overhead of a multi-statement lambda. I typically use computeIfAbsent if the lambda can fit on a single line.

So to answer your question: I didn't implement it for stylistic reasons.

But you did bring up a good point the I'm using a `ConcurrentHashMap` which is overkill in this particular case, so I'll change it to a `HashMap`.

File: bundles/org.openhab.binding.bluetooth.bluegiga/src/main/resources/ESH-INF/thing/bluegiga.xml (hunk 1 above)
Comment (cpmeister) [reply bundles/org.openhab.binding.bluetooth.bluegiga/src/main/resources/ESH-INF/thing/bluegiga.xml]: done

File: bundles/org.openhab.binding.bluetooth.bluegiga/src/main/resources/ESH-INF/thing/bluegiga.xml (hunk 2 above)
Comment (cpmeister) [reply bundles/org.openhab.binding.bluetooth.bluegiga/src/main/resources/ESH-INF/thing/bluegiga.xml]: done

File: bundles/org.openhab.binding.bluetooth.bluegiga/src/main/resources/ESH-INF/thing/bluegiga.xml (hunk 4 above)
Comment (cpmeister) [reply bundles/org.openhab.binding.bluetooth.bluegiga/src/main/resources/ESH-INF/thing/bluegiga.xml]: done

File: bundles/org.openhab.binding.bluetooth.bluez/src/main/resources/ESH-INF/thing/bluez.xml (hunk 8 above)
Comment (cpmeister) [reply bundles/org.openhab.binding.bluetooth.bluez/src/main/resources/ESH-INF/thing/bluez.xml]: done

File: bundles/org.openhab.binding.bluetooth.bluez/src/main/resources/ESH-INF/thing/bluez.xml (hunk 7 above)
Comment (cpmeister) [reply bundles/org.openhab.binding.bluetooth.bluez/src/main/resources/ESH-INF/thing/bluez.xml]: done

File: bundles/org.openhab.binding.bluetooth.bluegiga/src/main/resources/ESH-INF/thing/bluegiga.xml (hunk 3 above)
Comment (cpmeister) [reply bundles/org.openhab.binding.bluetooth.bluegiga/src/main/resources/ESH-INF/thing/bluegiga.xml]: done

File: bundles/org.openhab.binding.bluetooth.bluez/src/main/java/org/openhab/binding/bluetooth/bluez/handler/BlueZBridgeHandler.java (hunk 6 above)
Comment (cpmeister) [reply bundles/org.openhab.binding.bluetooth.bluez/src/main/java/org/openhab/binding/bluetooth/bluez/handler/BlueZBridgeHandler.java]: done

File: bundles/org.openhab.binding.bluetooth.bluez/src/main/java/org/openhab/binding/bluetooth/bluez/handler/BlueZBridgeHandler.java (hunk 5 above)
Comment (cpmeister) [reply bundles/org.openhab.binding.bluetooth.bluez/src/main/java/org/openhab/binding/bluetooth/bluez/handler/BlueZBridgeHandler.java:135]: I was surprised when I first found out about that too.
Using labels allows you to break out of any code block.
I rarely find the need to use it though but it does come in handy on occasion.

File: bundles/org.openhab.binding.bluetooth.bluez/src/main/resources/ESH-INF/thing/bluez.xml [hunk 10]
```diff
@@ -20,18 +20,18 @@
...
-				<advanced>true</advanced>
-				<default>60</default>
-			</parameter>
-			<parameter name="inactiveDeviceCleanupThresholdSecs" type="integer" min="1">
-				<label>Device Cleanup Threshold</label>
-				<description>The number of seconds a device can remain radio silent before it is eligible for cleanup</description>
-				<advanced>true</advanced>
-				<default>300</default>
-			</parameter>
+			<parameter name="inactiveDeviceCleanupInterval" type="integer" min="1" unit="s">
```
Comment (J-N-K) [inline bundles/org.openhab.binding.bluetooth.bluez/src/main/resources/ESH-INF/thing/bluez.xml:23]: formatting looks wrong

Comment (cpmeister) [reply bundles/org.openhab.binding.bluetooth.bluez/src/main/resources/ESH-INF/thing/bluez.xml:23]: Where?

Comment (cpmeister) [reply bundles/org.openhab.binding.bluetooth.bluez/src/main/resources/ESH-INF/thing/bluez.xml:23]: Ah, you must be looking at the incremental commits. The formatting was fixed in the followup merge commit.

Comment (cpmeister) [reply bundles/org.openhab.binding.bluetooth.bluez/src/main/resources/ESH-INF/thing/bluez.xml:23]: Ah, you must be looking at the incremental commits. The formatting was fixed in the followup merge commit.

**Pull request merged**